### DISCUSSION TITLE
Single Fetch action/loader types with ResponseStub

### DISCUSSION THREAD

**Discussion by ksmithut** [remix-run/remix#9326] [Q&A]
I've been playing around with a new Remix app that has enabled the newly available (unstable) [Single Fetch](https://remix.run/docs/en/main/guides/single-fetch) feature. One of the new paradigm shifts is that instead of returning a response, it's recommended to just return data and use the `response` object that comes into the parameters when you need to change the status code or add headers.

I might be doing something wrong, but the types for `ActionFunction` and `LoaderFunction` both have `response` being `ResponseStub | undefined`. My understanding is that if I'm using "Single Fetch", the `response` should always be there. Is there some utility types I should be using? Or is it enough to just do `ActionFunctionArgs & { response: ResponseStub }`?

**Comment by kiliman** [comment]
The reason is to prevent a breaking change. Since this is a _future_ API that you opt in, they can't change the function signature directly. 

They make note of it here:

https://github.com/remix-run/remix/blob/7b2fa032cd465c7f6e9c8d90471d6a544e0c08f6/packages/remix-server-runtime/routeModules.ts#L95

However they could do something similar to `useLoaderData` types by adding future types to _tsconfig.json_

> You must add node_modules/@remix-run/react/future/single-fetch.d.ts to the end of your tsconfig.json's include array to get proper type inference 

You can also create your own types that will override the existing ones.

**Reply by brophdawg11** [reply]
It's worth noting that even with single fetch enabled, `response` is not provided to resource route loaders because they're just API endpoints and you return the `Response` directly - there's no need for a stub which is later merged with other stubs to generate a final `Response`.

Unsure what the best solution is there, of the options I can think off off the top of my head ($) is probably my favorite:

1. leave it as is
2. introduce a `ResourceRouteLoaderFunctionArgs` type to help differentiate
3. force users to use the `response` stub
4. pass it and respect it if it's used - but recommend against using it in favor of just setting status/headers on your returned `Response` instance

**Reply by brophdawg11** [reply]
Although under the hood that's a bit at odds with the implementation.  Response stubs are a feature of data strategy, data strategy is only a thing when you are querying _multiple_ routes (via `staticHandler.query`) and not a single direct loader/action (`staticHandler.queryRoute`), and resource routes use `queryRoute` so they have no access to a `dataStrategy` at the moment...

**Reply by alexanderson1993** [reply]
I think there was a suggestion at one point to make resource routes export something entirely different from `loader` and `action` - `export const resource` or `export const handler`, so they behave more like a proper request/response handler, handling all HTTP verbs, and less like a data strategy implementation.

https://github.com/remix-run/remix/discussions/7640#discussioncomment-7502465

**Reply by brophdawg11** [reply]
This would effectively be the same as (2) above, right?  Regardless of the name of the export, we'd have to introduce a new `ResourceRouteLoaderFunctionArgs` type:

```ts
type LoaderFunctionArgs = {
  request: Request,
  params: Params,
  context: AppContext,
  response: ResponseStub
}

type ResourceRouteLoaderFunctionArgs = {
  request: Request,
  params: Params,
  context: AppContext,
}
```

**Reply by alexanderson1993** [reply]
Ah, yes. That is true.

**Reply by maranomynet** [reply]
@brophdawg11 what about:
 
5. Pass the `response` stub, but ignore it and if it's used, log a warning in dev mode

That way you effectivly maintain the current behavior, but get away with simpler, more robust types for page loader functions (most common use case)

**Reply by brophdawg11** [reply]
we want it to be able to be used so folks can share utilities functions between resource and non-resource route loaders
